module/purpose/Soccer/src/Soccer.cpp

```diff
@@ -124,8 +162,131 @@ namespace module::purpose {
                 log<NUClear::INFO>("Force playing started.");
                 cfg.force_playing = true;
                 emit<Task>(std::make_unique<FindPurpose>(), 1);
+                auto msg = std::make_unique<RoboCup>();
+                msg->current_pose.player_id = PLAYER_ID;
+                emit(msg);
+            }
+        });
+
+        // check if we have heard from robots periodically, and remove them if not
+        on<Every<2, Per<std::chrono::seconds>>>().then([this] {
+            auto now = std::chrono::steady_clock::now();
+            constexpr auto timeout = std::chrono::seconds(8); //TODO: test with robot or get advice
+
+            // remove inactive robots
+            for (auto it = active_robots.begin(); it != active_robots.end(); ) {
+                if (it->robot_id != PLAYER_ID && now - it->last_heard_from > timeout) {
+                    log<NUClear::DEBUG>("remove player ", int(it->robot_id));
+                    it = active_robots.erase(it);
+                    give_directions();
+                } else {
+                    ++it;
+                }
             }
+
+            log<NUClear::DEBUG>("active robots length ", int(active_robots.size()));
+        });
+    }
+
+    void Soccer::add_robot(RobotInfo new_robot) {
+        // Find the correct insertion point
+        auto insertPos = std::lower_bound(active_robots.begin(), active_robots.end(), new_robot,
+            [](const RobotInfo& a, const RobotInfo& b) {
+                return a.robot_id < b.robot_id;
+            });
+
+        // Insert the new robot at the found position to maintain order
+        active_robots.insert(insertPos, new_robot);
+    }
+
+    void Soccer::manage_active_robots(const RoboCup& robocup) {
+        // do not manage penalised robots
+        if (robocup.state == State::PENALISED) { return; }
+
+        uint8_t incoming_robot_id = robocup.current_pose.player_id;
+        auto now = std::chrono::steady_clock::now();
+
+        std::vector<RobotInfo>::iterator it = std::find_if(active_robots.begin(), active_robots.end(), [incoming_robot_id](const RobotInfo& info) {
+            return info.robot_id == incoming_robot_id;
         });
+
+        if (it != active_robots.end()) {
+            // if the robot is in active_robots, update timestamp
+            it->last_heard_from = now;
+            // add more data here
+        } else {
+            // Add new robot's info, and re-assign positions
+            RobotInfo new_robot{incoming_robot_id, now};
+            add_robot(new_robot);
+            give_directions();
+        }
+
+        follow_directions(robocup, incoming_robot_id);
+    };
+
+    // find which robot should be our striker
+    uint8_t Soccer::find_striker() {
+        return 0;
+        // TODO: ideas- furthest ahead on field could be striker, or closest to ball
```

Comment:
also want to keep in mind that we don't want the striker changing when it doesn't need to. 
ie player 1 is penalised, player 2 is striker
player 2 is close to scoring
player 1 is unpenalised and becomes the striker
player 2 moves back to defense when it could have scored

Reply:
it might be good to get this merged though then add this on, as this PR will work well for single-robot scenarios and it might be a bit more work to add in this functionality.